consensus/safety-rules/src/safety_rules.rs

```diff
@@ -195,32 +190,37 @@ impl TSafetyRules for SafetyRules {
 
     fn initialize(&mut self, proof: &EpochChangeProof) -> Result<(), Error> {
         let waypoint = self.persistent_storage.waypoint()?;
+        let verifier: &dyn Verifier = match self.epoch_state() {
+            Ok(epoch_state) => epoch_state,
+            Err(_) => &waypoint,
+        };
         let last_li = proof
-            .verify(&waypoint)
-            .map_err(|e| Error::WaypointMismatch(format!("{}", e)))?;
+            .verify(verifier)
+            .map_err(|e| Error::InvalidEpochChangeProof(format!("{}", e)))?;
         self.start_new_epoch(last_li.ledger_info())
     }
 
-    /// Verify the QC is correct and up to date, if it is either set the preferred round or start a
-    /// new epoch.
-    fn update(&mut self, qc: &QuorumCert) -> Result<(), Error> {
-        self.verify_qc(qc)?;
-        if qc.ends_epoch() {
-            self.start_new_epoch(qc.ledger_info().ledger_info())
-        } else {
-            self.persistent_storage
-                .set_preferred_round(qc.parent_block().round())
-                .map_err(|e| e.into())
-        }
-    }
-
     /// @TODO verify signature on vote proposal
-    /// @TODO verify QC correctness
     fn construct_and_sign_vote(&mut self, vote_proposal: &VoteProposal) -> Result<Vote, Error> {
         debug!("Incoming vote proposal to sign.");
-        self.signer()?;
         let proposed_block = vote_proposal.block();
         self.verify_epoch(proposed_block.epoch())?;
+        self.verify_qc(proposed_block.quorum_cert())?;
+        let preferred_round = self.persistent_storage.preferred_round()?;
+        if proposed_block.quorum_cert().parent_block().round() > preferred_round {
+            self.persistent_storage
+                .set_preferred_round(proposed_block.quorum_cert().parent_block().round())?;
+        }
+        if proposed_block.quorum_cert().certified_block().round() < preferred_round {
+            debug!(
+                "Vote proposal certified round is lower than preferred round, {} < {}",
+                proposed_block.quorum_cert().certified_block().round(),
+                preferred_round,
+            );
+            return Err(Error::ProposalRoundLowerThenPreferredBlock { preferred_round });
+        }
```

Comment:
Recommend we do error checking first. It would be pretty weird that we update our preferred round and then error our because the certified round is older :O

Also as I mentioned in the other pr, this should just go into verify_qc replacing what you have there.

Reply:
also could we add some white space between the if statements?

Reply:
one last ask, could we add a log statement for the case where we find an older parent block then preferred round? it is an unexpected and uncommon event, so it would be good to note that we're in such a state if we have issues getting blocks committed.